### PR TITLE
Refactored &  fixed NoteRead feature 

### LINKED ISSUE #60: Implement Text-to-Speech Feature for Note Content Reading
In the NotesReadOnlyPage, we can introduce a "Play" button that, when clicked, will read aloud the contents of the note, including both the title and the body.

To accomplish this, we can leverage the [flutter_tts](https://pub.dev/packages/flutter_tts) library, which provides text-to-speech capabilities for Flutter applications.

Additionally, as part of this enhancement, we can explore the possibility of enabling background playback for the audio. This means that users can continue to listen to their notes being read even when the app is not in focus or running in the background.

For now we can use hardcoded voice type, we can make voice, pitch and other things configurable in future.

 

Body of the note will be in delta form, it can be converted to plainText by `state.controller!.document.toPlainText()`.

### PR DESCRIPTION
**Pull request by vaibhavppandey**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
refactored the code made by the prev contributor and fixed minor bugs 

## Features Covered in this PR

- [x] made a different `NoteReadIconButton` widget to seperate TTS feat 
- [x] fixed bug, where the icon changes to indicate play when TTS completion happens

## Related Tickets & Documents

- Related Issue #60 
- Closes #60 

## Screenshots, Recordings

https://github.com/SankethBK/diaryvault/assets/113239561/a4d5aad1-2961-4a49-bff9-ce07c3e5c348

## Tested Feature??

- [x] In Real Device.
- [x] In Emulator

### CODE REVIEW

**Comment by vaibhavppandey** [comment]
Also @SankethBK, I was planning to implement the `bloc` arch for state management for this feat. But given the feat's current scope , I went with the simple `setState` .  But I am ready to implement the `bloc` arch if you say so.

**Comment by SankethBK** [comment]
@vaibhavppandey `setState` is sufficient for this, no need for bloc as the state can be contained within the widget

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      });
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final notesBloc = BlocProvider.of<NotesBloc>(context);
+    return IconButton(
+      icon: Icon(isPlaying ? Icons.pause : Icons.play_arrow),
+      onPressed: () async {
```
Comment (SankethBK) [inline lib/features/notes/presentation/widgets/note_read_button.dart:54]: Can you wrap this with
```
Padding(
      padding: const EdgeInsets.only(right: 13.0),
```

just like we have for `NoteSaveButton`

File: lib/features/notes/presentation/pages/note_read_only_page.dart [hunk 2]
```diff
@@ -35,11 +38,12 @@ class _NotesReadOnlyPageState extends State<NotesReadOnlyPage> {
...
+  // note reading
+  bool isPlaying = false;
+  bool isPaused = false;
+  late FlutterTts flutterTts;
+
 
-  @override
-  void initState() {
-    super.initState();
-  }
```
Comment (SankethBK) [inline lib/features/notes/presentation/pages/note_read_only_page.dart:47]: I assume these are previous changes from this branch, can you revert this part

File: lib/features/notes/presentation/pages/note_read_only_page.dart [hunk 3]
```diff
@@ -227,6 +232,11 @@ class _NotesReadOnlyPageState extends State<NotesReadOnlyPage> {
     );
   }
 
+  Future<void> _speak(String text) async {
+    await flutterTts.setLanguage("en-US"); // Set the language if needed
+    await flutterTts.setPitch(1.0); // Set pitch (1.0 is the default)
+    await flutterTts.speak(text);
+  }
```
Comment (SankethBK) [inline lib/features/notes/presentation/pages/note_read_only_page.dart:239]: I assume these are previous changes from this branch, can you revert this part

**Comment by SankethBK** [comment]
Minor comments, rest all looks good! Nice job!!

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      setState(() {
+        isPlaying = true;
+      });
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final notesBloc = BlocProvider.of<NotesBloc>(context);
+    return Padding(
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:46]: Please add a white line before the return statement.

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ..setLanguage("en-US")
+    ..setSpeechRate(0.5)
+    ..setPitch(1.0);
+  bool isPlaying = false;
+
+  @override
+  void initState() {
+    super.initState();
+    flutterTts.setCompletionHandler(() {
+      setState(() {
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:25]: It's not best practice to use `setState(); `
Please use [`ValueNotifier`](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html) & [`ValueListenableBuilder`](https://api.flutter.dev/flutter/widgets/ValueListenableBuilder-class.html) to achive the same result... Thank you so much!!

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        isPlaying = false;
+      });
+    });
+
+    flutterTts.setPauseHandler(() {
+      setState(() {
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:31]: [Same as above](https://github.com/SankethBK/diaryvault/pull/65/files#r1349370717)

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        isPlaying = false;
+      });
+    });
+
+    flutterTts.setContinueHandler(() {
+      setState(() {
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:37]: [Same here](https://github.com/SankethBK/diaryvault/pull/65/files#r1349370717)

**Review by micedreams — COMMENTED** [review]
Your PR looks pretty good except for a few things...good job!!

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        icon: Icon(isPlaying ? Icons.pause : Icons.play_arrow),
+        onPressed: () async {
+          if (isPlaying) {
+            await flutterTts.pause();
+          } else {
+            await _speak(notesBloc.state.title! +
+                ". " +
+                notesBloc.state.controller!.document.toPlainText());
+          }
+          setState(() => isPlaying = !isPlaying);
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:58]: [here too](https://github.com/SankethBK/diaryvault/pull/65/files#r1349370717)

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:flutter_tts/flutter_tts.dart';
+
+
+class NoteReadIconButton extends StatefulWidget {
+  const NoteReadIconButton({Key? key}) : super(key: key);
+
+  @override
+  State<NoteReadIconButton> createState() => _NoteReadIconButtonState();
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:11]:
```suggestion
  State createState() => _NoteReadIconButtonState();
```

File: lib/features/notes/presentation/widgets/note_read_button.dart [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+class _NoteReadIconButtonState extends State<NoteReadIconButton> {
+  final FlutterTts flutterTts = FlutterTts()
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:15]:
```suggestion
  final flutterTts = FlutterTts()
```

**Comment by micedreams** [comment]
> @vaibhavppandey `setState` is sufficient for this, no need for bloc as the state can be contained within the widget

@SankethBK I just saw this... please take it back!!😂😂 `SetState` is never sufficient...

File: lib/features/notes/presentation/widgets/note_read_button.dart (hunk 4 above)
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_read_button.dart:46]:
```suggestion

    return Padding(
```

**Comment by SankethBK** [comment]
> > @vaibhavppandey `setState` is sufficient for this, no need for bloc as the state can be contained within the widget
> 
> @SankethBK I just saw this... please take it back!!😂😂 `SetState` is never sufficient... If you don't agree with me [this](https://articles.wesionary.team/why-you-should-opt-for-better-state-management-instead-of-using-setstate-in-flutter-b7fff1dac247#:~:text=Flutter's%20reactive%20nature%20can%20cause,the%20UI%20needs%20an%20update.) might help change your mind...

I agree, it is definitely not a good practice to call `setState` on a widget that has large number of children widgets. But since this widget doesn't has any child widget, i think it should be fine

**Comment by vaibhavppandey** [comment]
> > @vaibhavppandey `setState` is sufficient for this, no need for bloc as the state can be contained within the widget
> 
> @SankethBK I just saw this... please take it back!!😂😂  
> `SetState` is never sufficient...
> If you don't agree with me [this](https://articles.wesionary.team/why-you-should-opt-for-better-state-management-instead-of-using-setstate-in-flutter-b7fff1dac247#:~:text=Flutter's%20reactive%20nature%20can%20cause,the%20UI%20needs%20an%20update.) might help change your mind...
> 

Ya I second this but as @SankethBK mentioned, the `NoteReadIconButton` isn't complex yet so I used `setState`. However,  based on the feedback , I've transitioned to`ValueNotifier`. Thank you for the review! Got to learn something new

**Review by SankethBK — APPROVED** [review]
LGTM, thanks for contributing!

**Comment by SankethBK** [comment]
@micedreams I don't see any new linter errors, once we fix the ones in master we'll be able to track it easily.

**Pull request merged**